[css-anchor-position-1] anchor-name mutation doesn't propagate to its anchor-positioned dependency

anttijk opened the pull request:
#### 16977f92e64ac17ddc03794b2eb269636b659855
<pre>
[css-anchor-position-1] anchor-name mutation doesn&apos;t propagate to its anchor-positioned dependency
<a href="https://bugs.webkit.org/show_bug.cgi?id=294129">https://bugs.webkit.org/show_bug.cgi?id=294129</a>
<a href="https://rdar.apple.com/152727401">rdar://152727401</a>

Reviewed by Alan Baradlay.

We failed to invalidate the anchored elements in some cases when an anchor changed its name.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-mutation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-mutation.html: Added.

Test by Kiet.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Include anchor references that we fail to resolve into anchorPositionedToAnchorMap. They may get resolved later.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Remove a hack where we created anchor positioned state for anything with position-anchor here. It was hiding invalidation bugs.

(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

No resolved anchor -&gt; reset scroll offset.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Collect any changed anchor names and scopes whens style changes.

(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):

If there are changed anchor names use them to invalidate any existing anchor positioned elements for another round of interleaving.
Ensure we do at least one round of interleaving if there are any style changes and any anchor positioned elements as any style
change may cause anchors to move.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Remove setting of m_needsInterleavedLayout on seeing any anchor names, it is no longer needed and was hiding bugs.

(WebCore::Style::TreeResolver::collectChangedAnchorNames):

Collect the changed names and names affected by changes in anchor scopes.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/296184@main">https://commits.webkit.org/296184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b1db59c46b7979b7bae8d596df6252c3563f499

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17715 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112837 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58162 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109583 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27994 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81716 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62096 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57600 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91564 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15177 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115940 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34689 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90750 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90491 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23074 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35407 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13193 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30451 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34608 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40164 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37715 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36017 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tuankiet65 commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -1019,6 +1019,9 @@ void AnchorPositionEvaluator::updateSnapshottedScrollOffsets(Document& document)
                 return false;
 
+            if (!elementAndAnchors.value[0].renderer)
+                return false;
```
Maybe there could be two elements in `elementAndAnchors.value` and the second one is non null?

anttijk replied:
There is a test just above that there is exactly one element there.

 (this stuff is likely at least somewhat wrong)

tuankiet65 replied:
Right, what I mean is in that case we should've selected the second element. But tbf it's already wrong-ish so doesn't really make it worst.

tuankiet65 commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1628,6 +1646,44 @@ bool TreeResolver::hasResolvedAnchorPosition(const Styleable& styleable) const
+    // A changed anchor name is either a name being added, a name being removed, or a name whose interpretation changes.
+    // This may change which elements get anchored to it.
+
+    if (!currentStyle || currentStyle->anchorNames() != newStyle.anchorNames()) {
```
What if they have the same names but in different order (probably doesn't matter that much since it doesn't hurt to redo matching)

anttijk replied:
Yeah, we do invalidate it in that case too, unnecessarily. There is a comment below that mentions we could do more checking.

anttijk committed:
[css-anchor-position-1] anchor-name mutation doesn't propagate to its anchor-positioned dependency
https://bugs.webkit.org/show_bug.cgi?id=294129
rdar://152727401

Reviewed by Alan Baradlay.

We failed to invalidate the anchored elements in some cases when an anchor changed its name.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-mutation-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-mutation.html: Added.

Test by Kiet.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Include anchor references that we fail to resolve into anchorPositionedToAnchorMap. They may get resolved later.

(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):

Remove a hack where we created anchor positioned state for anything with position-anchor here. It was hiding invalidation bugs.

(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

No resolved anchor -> reset scroll offset.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Collect any changed anchor names and scopes whens style changes.

(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):

If there are changed anchor names use them to invalidate any existing anchor positioned elements for another round of interleaving.
Ensure we do at least one round of interleaving if there are any style changes and any anchor positioned elements as any style
change may cause anchors to move.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Remove setting of m_needsInterleavedLayout on seeing any anchor names, it is no longer needed and was hiding bugs.

(WebCore::Style::TreeResolver::collectChangedAnchorNames):

Collect the changed names and names affected by changes in anchor scopes.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/296184@main
